### DISCUSSION TITLE
Application use a lot of memory with Bucket4j and Springboot 3

### DISCUSSION THREAD

**Discussion by williamjvr** [vladimir-bukhtoyarov/bucket4j#486] [Q&A] [answered]
Hi everyone, I need a help with one problem, my application used a lot of memory and stoped after the throughput is increase, my application use Springboot3 ,Bucket4j_jdk17 and Java 21, below is a example of class to configure and create a bucket, another detail is i used Elasticache in AWS.

The flow of application is, every request received , validate the person is a token in bucket and after execute method, but my ideal throughput is 10K per minute.

All classes below is injected by Spring

<img width="854" alt="image" src="https://github.com/bucket4j/bucket4j/assets/2123490/294f95b9-61bd-4c55-a872-3d323f7aafcd">

<img width="849" alt="image" src="https://github.com/bucket4j/bucket4j/assets/2123490/96620fae-5a27-40ed-aac9-294811aa2bc5">
<img width="881" alt="image" src="https://github.com/bucket4j/bucket4j/assets/2123490/a0f1086c-60ad-4310-8f89-788f64bc4e07">

<img width="694" alt="image" src="https://github.com/bucket4j/bucket4j/assets/2123490/057f789b-5280-4dbf-858a-5dd478f730d2">

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @williamjvr 

You should do primary diagnostic by yourself. Configure JVM option to make heapdump on OOM. Or acquire heapdump manually. Then analyze heapdump in Eclipse Memory Analizer Tool(MAT). Without heapdump it is hard to answer where you were wrong. 

A few notes about your code:
1. You create one Redis connection per Bucket and one ProxyManager per bucket, it is waste of resources. Typically you should use only one proxyManager per application. Just create a bean fofor proxyManager. 
2. It is not nessesary to cache the buckets, if amount of buckets can not be predictable. Just create, use and forget bucket per each request. 

I make a bet that MAT will show that cause of leak is connection per bucket problem.

**Reply by williamjvr** [reply]
Thanks for this feeedback @vladimir-bukhtoyarov , I will try to configure one proxyManager per application, and i analyzed do heapdump and below a image with points to problem

<img width="615" alt="image" src="https://github.com/bucket4j/bucket4j/assets/2123490/bbbaa187-9aaf-45e9-97e4-fe16fbec8c36">

**Reply by vladimir-bukhtoyarov** [reply]
Dump shows that heap was consumed by Redis connections. Thereason and solution are described above.
